source/server/admin/admin_no_html.cc

```diff
@@ -1,14 +1,41 @@
-#include "source/server/admin/admin.h"
+#include "envoy/server/admin_handler_factory.h"
 
 namespace Envoy {
 namespace Server {
 
-Http::Code AdminImpl::handlerAdminHome(Http::ResponseHeaderMap& response_headers,
-                                       Buffer::Instance& response, AdminStream&) {
-  response_headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Text);
-  response.add("HTML output was disabled by building with --define=admin_html=disabled");
-  return Http::Code::OK;
-}
+class AdminHandlerHome : public AdminHandler {
+public:
+  const std::string& prefix() const override {
+    static const std::string prefix = "/";
+    return prefix;
+  }
+  const std::string& helpText() const override {
+    static const std::string help_text = "Admin home page";
+    return help_text;
+  }
+  bool removable() const override { return false; }
+  bool mutatesServerState() const override { return false; }
+  const Admin::ParamDescriptorVec& params() const override {
+    static const Admin::ParamDescriptorVec& params{};
+    return params;
+  }
+  Http::Code runCallback(Http::ResponseHeaderMap& response_headers, Buffer::Instance& response,
+                         AdminStream&) override {
+    response_headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Text);
+    response.add("HTML output was disabled by building with --define=admin_html=disabled");
+    return Http::Code::OK;
+  }
+};
+
+class AdminHandlerHomeFactory : public AdminHandlerFactory {
+public:
+  std::unique_ptr<AdminHandler> createHandler(Admin&, Server::Instance&) const override {
+    return std::make_unique<AdminHandlerHome>();
+  }
+  std::string name() const override { return "home"; }
+};
+
+REGISTER_FACTORY(AdminHandlerHomeFactory, AdminHandlerFactory);
```

Comment:
I personally am not a fan of this registration pattern we use but I understand why it makes the code simpler and others like it, and it's fine with me if that's what people want.

Every one of these takes us further from "envoy as a library", which I think is a concept that Envoy-Mobile is trying to go to, which I also think makes testing cleaner. Having said this I'm ready to move on if others like this.